enhancements/update/cvo-techpreview-manifests.md

```diff
@@ -0,0 +1,139 @@
+is the canonical way to enable tech preview features.
+It is the setting that all operators use to enable Tech Preview features.
+The CVO will honor this setting on a per-manifest basis, so that tech preview manifests are never present on
+non-tech preview clusters.  
+
+## Motivation
+
+When introducing tech preview operators, the current state of the art (before this enhancement) is to have the 
+operator build an inert mode that detects if tech preview is not enable and then does not take action.
+This approach results in having additional resources, including the clusteroperator/tech-preview itself, 
+installed on every cluster, including non-tech-preview upgrades. 
+This also means that currently, if the tech preview does not progress to stable, the component has to keep "dead" manifests
+around and marked for deletion by the CVO on upgraded clusters.
+By having the CVO honor tech preview per manifest, we can avoid fanning out inert logic to every tech preview operator,
+we can avoid exposing tech preview operators to customers via clusteroperators, and we can avoid creating unnecessary
+resources like namespaces, roles, rolebindings, serviceacounts, and deployments.
+
+### Goals
+
+1. CVO honors an "only create or update this manifest in clusters with tech preview enabled".
+
+### Non-Goals
+
+1. Bootstrap rendering for tech preview operators.
+2. 4.y upgrades.  TechPreviewNoUpgrade explicitly disallows upgrades from 4.y to 4.y+1.
+3. Cleanup of manifests created as TechPreviewNoUpgrade.  TechPreviewNoUpgrade explicitly disallows being unset and this
+    is enforced using validation on the apiserver.
+4. Allow manifests to be removed or un-reconciled if TechPreviewNoUpgrade is *not* set.  There are a few case (mostly
+    around CRDs), where this capability is useful, but it is not as common as needing to create something additional.
+
+## Proposal
+
+Add an annotation that can be set on CVO manifests called `config.openshift.io/feature-gate` with one legal/honored value
+of "TechPreviewNoUpgrade".
+If the value is not "TechPreviewNoUpgrade", the manifest is never created.
+Since most people add manifests in order to see them applied, they will notice if their manifest is never created and
+the choice of annotation allows for potential future usage in the CVO consistent with feature gate handling used by
+other operators.
+
+If a manifest sets `.metadata.annotations["config.openshift.io/feature-gate"]="TechPreviewNoUpgrade"`, the manifest will
+not be created unless `featuregates.config.openshift.io|.spec.featureSet="TechPreviewNoUpgrade"`.
+If `featuregates.config.openshift.io|.spec.featureSet="TechPreviewNoUpgrade"`, then the manifest is reconciled as normal
+for whatever stage the CVO is in.
+The `featuregates.config.openshift.io|.spec.featureSet="TechPreviewNoUpgrade"` can be included in the installation
+manifests provided by a customer (we do this some CI scenarios as well), but there is no expected CVO change for bootstrapping.
```

Comment:
Customer-provided manifests go up through the installer to the bootstrap node and are applied by [cluster-bootstrap][1].  There's no CVO connection there.

The CVO-bootstrapping bit I'm concerned about is:

1. Bootstrap node comes up.
2. Bootstrap CVO launched in a static pod.
3. Bootstrap CVO tries to fetch FeatureGates, but because it's early, that fetch fails.
4. CVO assumes no gates, and skips `TechPreviewNoUpgrade` manifests.
5. Something locks up, because bootstrapping this particular cluster depended on a tech-preview feature like "include support for `$NEW_HARDWARE`".

I think we can resolve my concern via either:

a. Declare that there won't be any `TechPreviewNoUpgrade` manifests that are needed for bootstrapping, so locking up if someone tries to add something like that is acceptable CVO behavior.
b. Declare that we might have that sort of thing, and explain how the CVO is expected to handle it (e.g. installer sets an environment variable or somehow informs the static bootstrap CVO pod that for this particular install, `TechPreviewNoUpgrade` is enabled).

[1]: https://github.com/openshift/cluster-bootstrap

Reply:
> a. Declare that there won't be any `TechPreviewNoUpgrade` manifests that are needed for bootstrapping, so locking up if someone tries to add something like that is acceptable CVO behavior.

This is what I was trying to describe with "there is no expected CVO change for bootstrapping".  How can I phrase it more clearly?

Reply:
You ca[](url)

Reply:
You can drop:

> The `featuregates.config.openshift.io|.spec.featureSet="TechPreviewNoUpgrade"` can be included in the installation manifests provided by a customer...

Because while folks _could_ do that, it's completely orthogonal to this enhancement and would have no effect unless `cluster-bootstrap` or the installer grows similar handling.

Instead, say something like:

> During bootstrapping, the CVO will assume no feature sets are enabled until it can successfully retrieve `featuregates.config.openshift.io` from the Kubernetes API server.